### PR TITLE
replication metrics

### PR DESCRIPTION
**Pull request by jdndiaye**
[#42](https://github.com/medic/cht-app-monitoring-data-ingest/issues/42) I did:
- Replication success rate per day (probably a stacked bar graph of successes and failures)
- Reasons why replications are failing - Probably a table
- Total number of docs replicated

### CODE REVIEW

File: postgres/functions/get_replication_by_status.sql [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    sumsqr
+  FROM telemetry_docs_with_metric_blob
+  CROSS JOIN LATERAL jsonb_to_record(metric_values) AS (min decimal, max decimal, sum decimal, count bigint, sumsqr decimal)
+)
+
+SELECT
+  current_database() AS partner,
+  period_start,
+  SUM(count) filter(where metric LIKE ''%:medic:%:success'') AS replication_success_count,
+  SUM(count) filter(where metric LIKE ''%:medic:%:failure'') AS replication_failure_count,
```
Comment (kennsippell) [inline postgres/functions/get_replication_by_status.sql:78]: can we add another column here for `:failure:reason:error`?

File: postgres/functions/get_replication_by_status.sql [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    count,
+    sumsqr
+  FROM telemetry_docs_with_metric_blob
+  CROSS JOIN LATERAL jsonb_to_record(metric_values) AS (min decimal, max decimal, sum decimal, count bigint, sumsqr decimal)
+)
+
+SELECT
+  current_database() AS partner,
+  period_start,
+  SUM(count) filter(where metric LIKE ''%:medic:%:success'') AS replication_success_count,
```
Comment (kennsippell) [inline postgres/functions/get_replication_by_status.sql:77]: can we `coalesce(x,0)` each of these?

File: postgres/functions/get_replication_doc_count.sql [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    count,
+    sumsqr
+  FROM telemetry_docs_with_metric_blob
+  CROSS JOIN LATERAL jsonb_to_record(metric_values) AS (min decimal, max decimal, sum decimal, count bigint, sumsqr decimal)
+)
+
+SELECT
+  current_database() AS partner,
+  metric,
+  SUM(count) AS docs_count
```
Comment (kennsippell) [inline postgres/functions/get_replication_doc_count.sql:77]: instead of using telemetry - can we should this same data from `monitoring_docs` table in `dwh_impact`?  it would be in docs with type = 'monitoring` at `{couchdb,medic,doc_count}` or something similar

i think it will be a much faster simpler query which is the same result

File: postgres/functions/get_replication_failure_reasons.sql [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  CROSS JOIN LATERAL jsonb_to_record(metric_values) AS (min decimal, max decimal, sum decimal, count bigint, sumsqr decimal)
+)
+
+SELECT
+  current_database() AS partner,
+  metric,
+  SUM(count) AS count
+FROM telemetry_metrics
+WHERE period_start >= now() - ''30 days''::interval
+    and metric like ''replication:medic:%:failure:%''
```
Comment (kennsippell) [inline postgres/functions/get_replication_failure_reasons.sql:80]:
```suggestion
    and metric like ''replication:medic:%:failure:reason:%''
```

I think this is the same but makes readability easier.

**Review by kennsippell — CHANGES_REQUESTED** [review]
just one major feedback on your approach to doc_count

File: postgres/functions/get_replication_by_status.sql (hunk 1 above)
Comment (kennsippell) [reply postgres/functions/get_replication_by_status.sql:78]: if you look in [observations here](https://docs.google.com/spreadsheets/d/1nUyRra9fqOK0jrKn4WCqTZMcdXoGJTrfaZ-xWWLHgk4/edit#gid=617106475) i think we want 2 bars charts per day:  one for success/failure and a second for errors/denied. This would give us both

File: postgres/functions/get_replication_doc_count.sql (hunk 3 above)
Comment (jdndiaye) [reply postgres/functions/get_replication_doc_count.sql]: I didn't get the same result. But by using the `monitoring_docs `table, the query runs faster.

File: postgres/functions/get_replication_by_status.sql [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+SELECT
+  current_database() AS partner,
+  period_start,
+  COALESCE(SUM(count) filter(where metric LIKE ''%:medic:%:success''), 0) AS replication_success_count,
+  COALESCE(SUM(count) filter(where metric LIKE ''%:medic:%:failure''), 0) AS replication_failure_count,
+  COALESCE(SUM(count) filter(where metric LIKE ''%:medic:%:denied''), 0) AS replication_denied_count,
+  COALESCE(SUM(count) filter(where metric LIKE ''%:failure:reason:error''), 0) AS replication_error_count
+FROM telemetry_metrics
+WHERE
+  period_start >= now() - ''30 days''::interval
```
Comment (kennsippell) [inline postgres/functions/get_replication_by_status.sql:83]: image

I forgot other dashboards are doing 60 days. Should we do the same?

File: postgres/matviews/replication_docs.sql [hunk 6]
```diff
@@ -0,0 +1,16 @@
+CREATE MATERIALIZED VIEW public.app_monitoring_replication_docs
+TABLESPACE pg_default
+AS
+  SELECT
+    created,
+    partner_name,
+    COALESCE((doc #>> '{couchdb,medic,doc_count}')::int, 0) AS doc_count
```
Comment (kennsippell) [inline postgres/matviews/replication_docs.sql:7]: seems weird to make a matview for just one metric.. but it seems to be what we're doing elsewhere in this codebase. 

any thoughts on this? should we merging view with others like it (eg. `users_replication_limit.sql` or `sms_error_states.sql`)?  cc @eljhkrr @freddieptf

Comment (jdndiaye) [reply postgres/matviews/replication_docs.sql:7]: @kennsippell  Yes, you are absolutely right, we can group them in one matview.

File: postgres/functions/get_replication_failure_reasons.sql [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+  FROM telemetry_docs_with_metric_blob
+  CROSS JOIN LATERAL jsonb_to_record(metric_values) AS (min decimal, max decimal, sum decimal, count bigint, sumsqr decimal)
+)
+
+SELECT
+  current_database() AS partner,
+  metric,
+  SUM(count) AS count
+FROM telemetry_metrics
+WHERE period_start >= now() - ''30 days''::interval
```
[the remark this answers is not in the document]
Comment (jdndiaye) [reply postgres/functions/get_replication_failure_reasons.sql:79]: @kennsippell don't see how you can combine get_replication_by_status.sql and get_replication_failure_reasons.sql. the aggregations are not the same.

File: postgres/matviews/replication_docs.sql (hunk 6 above)
Comment (jdndiaye) [reply postgres/matviews/replication_docs.sql:7]: I can make that change in this PR. The query takes 866 ms (3ms fetch)

File: postgres/functions/get_replication_failure_reasons.sql (hunk 7 above)
[the remark this answers is not in the document]
Comment (kennsippell) [reply postgres/functions/get_replication_failure_reasons.sql]: yes. sorry

File: postgres/functions/get_replication_performance.sql [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+)
+select
+  current_database() AS partner,
+  metric,
+  period_start,
+  min(min),
+  sum(count),
+  sum(sum) / sum(count) as mean,
+  max(max),
+  percentile_disc(0.25) within group (order by COALESCE(sum, 0) / count) as first_quartile,
```
Comment (kennsippell) [inline postgres/functions/get_replication_performance.sql:79]: i'm not sure we're totally ready for performance metrics yet in app-monitoring - can we do it in another PR? now that we have looked at exploratory metrics - i think we should talk as a group about what was valuable and what was not.  once we agree on what to build we can revisit this. code looks solid though!

**Review by kennsippell — APPROVED** [review]

File: postgres/matviews/sms_error_and_users_docs_replication.sql [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+            SELECT url_id,
+                doc,
+                ROW_NUMBER() OVER (
+                    PARTITION BY url_id,
+                    date_trunc(
+                        'day',
+                        to_timestamp(
+                            (
+                                (doc::jsonb#>>'{date,current}')::bigint / 1000
+                            )::double precision
```
Comment (kennsippell) [inline postgres/matviews/sms_error_and_users_docs_replication.sql:57]: if there are two rows with the `same day` and with `doctype = 'monitoring'` in monitoring_docs I believe doing a join like [here](https://github.com/medic/cht-app-monitoring-data-ingest/pull/48/files#diff-25ccf03359502816d961ad7b0dc2467113f78864a88208e0172c5aab7d39b6a1R79) will result in duplicate results in the final output

i believe this ROW_NUMBER() and `rank = 1` code is intended to prevent that. 
instead of using the `WITH` statement is it possible to use what is here and just add to the select

File: postgres/functions/get_replication_by_status.sql [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+  CROSS JOIN LATERAL jsonb_to_record(metric_values) AS (min decimal, max decimal, sum decimal, count bigint, sumsqr decimal)
+)
+
+SELECT
+  current_database() AS partner,
+  period_start,
+  COALESCE(SUM(count) filter(where metric LIKE ''%:medic:%:success''), 0) AS replication_success_count,
+  COALESCE(SUM(count) filter(where metric LIKE ''%:medic:%:failure''), 0) AS replication_failure_count,
+  COALESCE(SUM(count) filter(where metric LIKE ''%:medic:%:denied''), 0) AS replication_denied_count,
+  COALESCE(SUM(count) filter(where metric LIKE ''%:failure:reason:error''), 0) AS replication_error_count
```
Comment (eljhkrr) [inline postgres/functions/get_replication_by_status.sql:80]: It would be great if the failure reasons were captured and surfaced in a widget

File: postgres/functions/get_replication_performance.sql (hunk 8 above)
Comment (eljhkrr) [reply postgres/functions/get_replication_performance.sql:79]: Added to the upcoming check-in agenda

File: postgres/matviews/sms_error_and_users_docs_replication.sql (hunk 9 above)
Comment (kennsippell) [reply postgres/matviews/sms_error_and_users_docs_replication.sql:57]: alternatively you could decide it isn't needed because superset forces you to specify how aggregation is done when there are duplicate values...

File: postgres/functions/get_replication_by_status.sql (hunk 10 above)
Comment (jdndiaye) [reply postgres/functions/get_replication_by_status.sql:80]: you can see it with `replication_failure_reasons`

File: postgres/matviews/sms_error_and_users_docs_replication.sql [hunk 11]
```diff
@@ -49,12 +56,31 @@ create materialized view app_monitoring_sms_error_states as (
...
+  SELECT
+    sms.partner_name,
+    sms.reported,
+    sms.due,
+    sms.muted,
+    sms.failed,
+    sms.delivered,
+    sms.scheduled,
+    COALESCE((doc #>> '{replication_limit, count}')::int, 0) AS "users past replication_limit",
+    COALESCE((doc #>> '{couchdb,medic,doc_count}')::int, 0) AS doc_count
```
Comment (kennsippell) [inline postgres/matviews/sms_error_and_users_docs_replication.sql:76]: sorry i missed this - can we show this as the "change in docs per day"?

**Review by kennsippell — CHANGES_REQUESTED** [review]
missed [this](https://github.com/medic/cht-app-monitoring-data-ingest/pull/48/files#r1118972591) sorry

File: postgres/matviews/sms_error_and_users_docs_replication.sql (hunk 11 above)
Comment (jdndiaye) [reply postgres/matviews/sms_error_and_users_docs_replication.sql:76]: Ok noted

File: postgres/matviews/sms_error_and_users_docs_replication.sql [hunk 12]
```diff
@@ -63,21 +63,40 @@ SELECT COALESCE(mu.partner_name, mu.url) AS partner_name,
...
+), delta_of_docs AS (
+    SELECT c1.url_id,
+        c1.reported,
+        "users past replication_limit",
+        c1.doc_count_today - c2.doc_count_yesterday AS dalta_of_doc
+    FROM (
+        SELECT
+            url_id,
+            TO_DATE(created::text, 'YYYY-MM-DD') as reported,
+            COALESCE((doc #>> '{replication_limit, count}')::int, 0) AS "users past replication_limit",
```
Comment (kennsippell) [inline postgres/matviews/sms_error_and_users_docs_replication.sql:75]: Did you consider using the `lead()` or `lag()` postgres functions for this?

File: postgres/functions/get_replication_by_status.sql (hunk 10 above)
Comment (eljhkrr) [reply postgres/functions/get_replication_by_status.sql:80]: My bad, I was expecting this field to have error detail but it only has a count https://docs.communityhealthtoolkit.org/apps/guides/performance/telemetry/

File: postgres/matviews/sms_error_and_users_docs_replication.sql [hunk 13]
```diff
@@ -0,0 +1,87 @@
+/* In this view we regroup the metrics of:
+  - sms_error_states
+  - users_replication_limit
+  - replication docs
+*/
+CREATE MATERIALIZED VIEW public.app_monitoring_sms_error_and_users_docs_replication
+TABLESPACE pg_default
+AS
+WITH sms_error_states AS (
```
Comment (kennsippell) [inline postgres/matviews/sms_error_and_users_docs_replication.sql:9]: you don't need to do this here - but bonus points if you do

i think the `WITH` and all the `ROW_NUMBER` stuff can be removed entirely with a single `DISTINCT ON` statement at the top

```
  select
    distinct on (partner_name, reported)
    ... selects from sms ... 
    ... your new selects ...
  FROM monitoring_docs AS docs
  INNER JOIN monitoring_urls AS urls ON (docs.url_id=urls.id)
  WHERE doctype = 'monitoring'
```

**Review by kennsippell — APPROVED** [review]
great change

**Comment by kennsippell** [comment]
Can you investigate why this isn't working for brac-ug before merging?

File: postgres/matviews/sms_error_and_users_docs_replication.sql [hunk 14]
```diff
@@ -6,79 +6,31 @@
...
+SELECT DISTINCT ON (partner_name, reported) partner_name,
+reported,
+due,
+muted,
+failed,
+delivered,
+scheduled,
+"users past replication_limit",
+delta_of_doc
+From(
```
Comment (kennsippell) [inline postgres/matviews/sms_error_and_users_docs_replication.sql:18]: `select 1 from (select 2)`

why do you need select 1?

Does this work?
```
SELECT 
  DISTINCT ON (partner_name, reported) 
  COALESCE(partner_name, urls.url) AS partner_name,
  TO_DATE(created::text, 'YYYY-MM-DD') AS reported,
  ...
```

File: postgres/matviews/sms_error_and_users_docs_replication.sql [hunk 15]
```diff
@@ -6,31 +6,20 @@
...
-        COALESCE((doc #>>'{messaging,outgoing,state,delivered}')::int,0)as delivered,
-        COALESCE((doc #>>'{messaging,outgoing,state,scheduled}')::int,0) as scheduled,
-        COALESCE((doc #>> '{replication_limit, count}')::int, 0) AS "users past replication_limit",
-        COALESCE((doc #>> '{couchdb,medic,doc_count}')::int, 0) - COALESCE(lag((doc #>> '{couchdb,medic,doc_count}')::int)
-            OVER (PARTITION BY docs.url_id ORDER BY created), 0) AS delta_of_doc
-    FROM monitoring_docs AS docs
-    INNER JOIN monitoring_urls AS urls ON (docs.url_id=urls.id)
-    WHERE doctype = 'monitoring'
-) AS ms
+SELECT distinct ON (COALESCE(partner_name, urls.url),
```
Comment (kennsippell) [inline postgres/matviews/sms_error_and_users_docs_replication.sql:9]:
```suggestion
SELECT DISTINCT ON (COALESCE(partner_name, urls.url),
```

**Review by kennsippell — APPROVED** [review]

File: postgres/matviews/sms_error_and_users_docs_replication.sql (hunk 14 above)
Comment (jdndiaye) [reply postgres/matviews/sms_error_and_users_docs_replication.sql]: > `select 1 from (select 2)`
> 
> why do you need select 1?
> 
> Does this work?
> 
> ```
> SELECT 
>   DISTINCT ON (partner_name, reported) 
>   COALESCE(partner_name, urls.url) AS partner_name,
>   TO_DATE(created::text, 'YYYY-MM-DD') AS reported,
>   ...
> ```

It works well. But I agree I can well delete the select 1.
